### PR TITLE
Openssl 1.1 support

### PR DESCRIPTION
**Pull request by xeroc**
This is an initial commit to include @nathanhourt's commit to support openssl-1.1.

However, it disables the blinding checks because they fail to compile. This needs to be fixed before merging.

### CODE REVIEW

**Comment by abitmore** [comment]
As @cwyyprog mentioned in https://github.com/bitshares/bitshares-core/issues/575, https://github.com/zaphoyd/websocketpp/issues/599 would be useful for this ticket.

**Comment by xeroc** [comment]
To fix this issue, I have placed a patched websocketpp into the bitshares project on github.
To apply the patch properly, you need to run

```
git submodule sync --recursive
git submodule update --init --recursive
```

//Edited by @abitmore: fixed typo in above commands.

**Comment by abitmore** [comment]
@xeroc: the websocketpp fix is included in `develop` branch of upstream repo with this commit: https://github.com/zaphoyd/websocketpp/commit/16d126ee61dfc901e75abc5573b704c72a8d1f24

**Comment by xeroc** [comment]
@abitmore I didn't want to move to the develop branch and instead used tag 0.7.0 + cherry-picked https://github.com/LocutusOfBorg/websocketpp/commit/1dd07113f2a7489444a8990a95be42e035f8e9df

**Comment by abitmore** [comment]
@xeroc: we've updated the websocketpp sub module with another PR, please rebase and/or fix conflicts. (Update: actually anyone who have write permission can do this, if have time)

**Comment by xeroc** [comment]
Rebased .. compiles here

**Comment by xeroc** [comment]
One of the unit tests fails ...
anyone capable that knows how to fix that?
@jmjatlanta ?

**Comment by jmjatlanta** [comment]
When I run blinding_test, I get ``Test setup error: test tree is empty``
When I run all_tests I get:

 ``unknown location(0): fatal error: in "fc_crypto/dh_test": memory access violation at address: 0x00000078: no mapping at fault address
/home/jmjatlanta/Development/cpp/bitshares-fc/tests/crypto/dh_test.cpp(11): last checkpoint
*** 1 failure is detected in the test module "AllTests"``

Is that what you get?
My Environment:
Boost 1.63.0 and OpenSSL 1.1.0i-dev

I also tested with OpenSSL 1.0.1m, with the same results. So it is not an OpenSSL 1.1 issue. It seems to be the way we are initializing the diffie hellman object within this test. The actual underlying openssl dh object is null. I am trying to find out how and when it should be initialized.

After further testing, it looks like https://github.com/bitshares/bitshares-fc/blob/cc26580c7cacf09792d1a99dca01d7c2e7de2af7/src/crypto/dh.cpp#L18 should be ``ssl_dh dh(DH_new());``

**Comment by xeroc** [comment]
Thanks, unit tests no longer fail!

//edit: they don't fail with openssl-1.0, but **do fail** with openssl-1.1 :-(

**Comment by pmconrad** [comment]
I think we should rip out the blinding stuff.

The core uses it only insofar as it provides access through the crypto_api. The implementation (and tests) has been broken since this commit: https://github.com/cryptonomex/fc/commit/2b2dfc6 , so I'm pretty sure nobody's using it.

**Comment by pmconrad** [comment]
> I think we should rip out the blinding stuff.

@abitmore what do you think?

**Comment by abitmore** [comment]
I agree.. I don't know much about crypto though.

**Comment by jmjatlanta** [comment]
kenCode was using some sort of blinding for his STEALTH coin. Should we ask him what he's relying on?

**Comment by pmconrad** [comment]
Like I said, the current implementation has been broken for >2 years. I'd be surprised if he's relying on that (and if he is he could as well maintain it himself...).

**Comment by xethyrion** [comment]
@pmconrad that implementation is partially relied upon and blind transfers do work using it so why rip out something that can benefit everyone?
-https://github.com/Agorise/bitshares-ui

**Comment by abitmore** [comment]
@xethyrion blind transfer in BitShares does not rely on that, but on another module.

By the way, linking the UI repository here doesn't make sense, since UI doesn't use fc library. Even if it does, please link the actually code block that uses it but not the main url of the repository.

**Comment by xethyrion** [comment]
@abitmore worried about witnesses which do need to validate blind tx's... chris should comment here on that.

**Comment by pmconrad** [comment]
Of course I do not propose to remove the existing STEALTH feature from the blockchain backend.

This is a different kind of blinding.  I implemented it myself in response to a bounty offered by BM, see https://bitsharestalk.org/index.php?topic=17315.0 .
The only reference to this in the core code is these two API calls: https://github.com/bitshares/bitshares-core/blob/master/libraries/app/api.cpp#L465-L477

**Comment by christophersanborn** [comment]
Ah, cool.  Thanks for the clarification @pmconrad!

**Comment by abitmore** [comment]
@pmconrad so we need to remove the 2 API calls if this got removed from FC?

**Comment by pmconrad** [comment]
yes

**Comment by pmconrad** [comment]
@christophersanborn thanks for confirming!

File: src/crypto/dh.cpp [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #include <fc/crypto/dh.hpp>
 #include <openssl/dh.h>
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
```
Comment (jmjatlanta) [inline src/crypto/dh.cpp:5]: Do these two lines do anything?

Comment (pmconrad) [reply src/crypto/dh.cpp:5]: No, that's probably a leftover C&P template or sth like that.

**Review by pmconrad — APPROVED** [review]
Tested with
* gcc-5, boost-1.60, openssl-1.0.2
* gcc-6, boost-1.65.1, openssl-1.0.2
* gcc-6, boost-1.65.1, openssl-1.1.0

With boost-1.65.1 the stacktrace unit tests fail, but I think that's unrelated.

All other known-working tests continue to work, replayed successfully + connected to mainnet.

**Comment by abitmore** [comment]
How about gcc-5, boost-1.58, openssl-1.0.2 (Ubuntu 16.04 LTS default packages)?

**Comment by pmconrad** [comment]
gcc-5, boost-1.58, openssl-1.0.2 tested successfully too.

**Pull request merged**